docs/labs/insecure-deserialization.html

```diff
@@ -0,0 +1,136 @@
+<!DOCTYPE html>
+<html>
+<head>
+<meta http-equiv="X-UA-Compatible" content="IE=edge">
+<meta name="viewport" content="width=device-width, initial-scale=1">
+<link rel="stylesheet" href="https://best.openssf.org/assets/css/style.css">
+<link rel="stylesheet" href="checker.css">
+<script src="js-yaml.min.js"></script>
+<script src="checker.js"></script>
+<link rel="license" href="https://creativecommons.org/licenses/by/4.0/">
+
+<!-- See create_labs.md for how to create your own lab! -->
+
+<!-- Sample expected answer -->
+<script id="expected0" type="plain/text">
+  const data = JSON.parse(base64Decoded);
+</script>
+<!--
+-->
+<script id="expected1" type="plain/text">
+  if (data.username && typeof data.username == 'string' && data.username.length < 20) {
+</script>
+
+<!-- Full pattern of correct answer -->
+<script id="correct0" type="plain/text">
+ const\s+data = JSON\.parse\(base64Decoded\)\; \s*
+</script>
+<script id="correct1" type="plain/text">
+ if \(data\.username && typeof\s+data\.username == ('string'|"string"|`string`) && data\.username\.length < 20\) \{ \s*
+</script>
+
+<script id="info" type="application/yaml">
+---
+hints:
+- present: "json"
+  text: the JSON built-in global object is witten in uppercase.
```

Comment:
Not everyone who takes the course will know JavaScript, and in particular, they may not be familiar with JavaScript syntax.

I suggest creating several hints for someone who's trying to fill in part1, but is getting stuck on one of the parts. Break down each part of the condition:

> data.username && typeof data.username == 'string' && data.username.length < 20) {

So if they're missing "data.username", suggest that as a way to detect if it's there. Then, if they're missing the typeof, suggest that. Many won't know if the result is 'String' or 'string', help them there. And so on. Adding a few hint patterns can really help someone who understands the concept but can't quite get the syntax right.

Reply:
Thanks for reviewing it! I'll add your suggestions.  

I tried to add more hints, but when I use "absent" for one of the parts, it still considers absent if the value is not in both parts. For example if I add "absent: length..." and then fill part 2 with length, the hint continues to show because length is not in part 1 as well. Am I doing something wrong?
Maybe is better if I change to have just one input area for both parts?

Reply:
By default it only looks at the first part, which is **index 0**. If you want it to look at a different part, you need to give it an **index: 1** or whatever the index is. You can see examples here:

* https://best.openssf.org/labs/sql-injection.html
* https://best.openssf.org/labs/regex1.html

If you want to check *both* places, currently you have to repeat the rule. If that happens a lot I guess we could add modify checker.js so you could provide `index: LIST`, but let's see if that happens often first...!